File: Tree/SearchBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def searchBST(self, root: Optional[TreeNode], v: int) -> Optional[TreeNode]:
        # ans = Optional[TreeNode]
        def search(root):
            if root is None:
                return None
            elif root.val==v:
                # ans = root
                return root
            elif root.val>v:
                return search(root.left)
            elif root.val<v:
                return search(root.right)
        result = search(root)
        return result

File: Tree/SearchBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* result;
    TreeNode* searchBST(TreeNode* root, int v) {
        //preorder traversal
        if(root==NULL){
            return NULL;
        }
        
        if(root->val==v){
            result=root;
            return result;
        }
        if(root->left && root->val>v){
            searchBST(root->left,v);
        }
        if(root->right && root->val<v){
            searchBST(root->right,v);
        }
        return result;
    }
};
